packages/core/streams/openai-stream.ts

```diff
@@ -117,10 +119,102 @@ function parseOpenAIStream(): (data: string) => string | void {
     return text
   }
 }
-
 export function OpenAIStream(
   res: Response,
   cb?: AIStreamCallbacks
 ): ReadableStream {
-  return AIStream(res, parseOpenAIStream(), cb)
+  const stream = AIStream(res, parseOpenAIStream(), cb)
+
+  if (cb && cb.onFunctionCall) {
+    const functionCallTransformer = createFunctionCallTransformer(cb)
+    return stream.pipeThrough(functionCallTransformer)
+  } else {
+    return stream
+  }
+}
+function createFunctionCallTransformer(
+  callbacks: AIStreamCallbacks
+): TransformStream<Uint8Array, Uint8Array> {
+  const textEncoder = new TextEncoder()
+  let isFirstChunk = true
+  let aggregatedResponse = ''
+  let isFunctionStreamingIn = false
+
+  return new TransformStream({
+    async transform(chunk, controller): Promise<void> {
+      const message = new TextDecoder().decode(chunk)
+      let newMessages: Message[] = []
+
+      const shouldHandleAsFunction =
+        isFirstChunk && message.startsWith('{"function_call":')
+      const isEndOfFunction =
+        !isFirstChunk &&
+        callbacks.onFunctionCall &&
+        isFunctionStreamingIn &&
+        message.endsWith('"}}')
+
+      if (shouldHandleAsFunction) {
+        isFunctionStreamingIn = true
+        aggregatedResponse += message
+        console.log('Function call detected')
+        isFirstChunk = false
+        return
+      }
+
+      // Stream as normal
+      if (!isFunctionStreamingIn) {
+        controller.enqueue(textEncoder.encode(message))
```

Comment:
```suggestion
        controller.enqueue(chunk)
```

To avoid doing the work again